src/models/NetworkAppController.js

```diff
@@ -169,18 +170,22 @@ export default class NetworkAppController {
     this.networkPackage.provider = { address: currentProvider.address };
   }
 
-  async uploadContracts() {
-    // TODO: Store the implementation's hash or full source/byte code to avoid unnecessary deployments
-    return Promise.all(_.map(this.package.contracts, async (contractName, contractAlias) => {
-      const contractClass = ContractsProvider.getFromArtifacts(contractName);
-      log.info(`Uploading ${contractName} implementation for ${contractAlias}`);
-      const contractInstance = await this.appManagerWrapper.setImplementation(contractClass, contractAlias);
-      log.info(`Uploaded ${contractName} at ${contractInstance.address}`);
-      this.networkPackage.contracts[contractAlias] = {
-        address: contractInstance.address,
-        bytecodeHash: bytecodeDigest(contractClass.bytecode)
-      };
-    }));
+  async uploadContracts(reupload) {
+    return Promise.all(
+      _(this.package.contracts)
+        .toPairs()
+        .filter(([contractAlias, contractName]) => reupload || this.hasContractChanged(contractAlias))
+        .map(async ([contractAlias, contractName]) => {
+          const contractClass = ContractsProvider.getFromArtifacts(contractName);
+          log.info(`Uploading ${contractName} implementation for ${contractAlias}`);
+          const contractInstance = await this.appManagerWrapper.setImplementation(contractClass, contractAlias);
+          log.info(`Uploaded ${contractName} at ${contractInstance.address}`);
+          this.networkPackage.contracts[contractAlias] = {
+            address: contractInstance.address,
+            bytecodeHash: bytecodeDigest(contractClass.bytecode)
+          };
+        }).value()
+    );
```

Comment:
let's extract the upload logic for a single contract to another function, do you agree?

Reply:
Sounds good!